### PR TITLE
feat: Node.js HTTP/2 Handler

### PR DESCRIPTION
**Pull request by trivikr**
*Issue #, if available:*
N/A

*Description of changes:*
Work in Progress, posting a Draft Pull Request for early look

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by trivikr** [comment]
The code in this PR is ready for review, unit tests are pending

File: packages/node-http-handler/src/node-http2-handler.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+} from "@aws-sdk/types";
+
+import { writeRequestBody } from "./write-request-body";
+import { getTransformedHeaders } from "./get-transformed-headers";
+
+export class NodeHttp2Handler
+  implements HttpHandler<Readable, NodeHttpOptions> {
+  private readonly connectionPool: Map<string, ClientHttp2Session>;
+
+  constructor(private readonly httpOptions: NodeHttpOptions = {}) {
```
Comment (AllanFly120) [inline packages/node-http-handler/src/node-http2-handler.ts]: Do we need a `NodeHttp2Options`? Is the Http1 and Http2 client option interfaces always the same?

File: packages/node-http-handler/src/node-http2-handler.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+          statusCode: headers[":status"] || -1,
+          headers: getTransformedHeaders(headers),
+          body: req
+        };
+        resolve(httpResponse);
+      });
+
+      req.on("error", reject);
+
+      const { connectionTimeout } = this.httpOptions;
```
Comment (AllanFly120) [inline packages/node-http-handler/src/node-http2-handler.ts]: Maybe it's out of the scope of the PR but maybe you have answer this: we support a per request httpOptions setting like [here](https://github.com/aws/aws-sdk-js-v3/blob/c4edc453aa1a6bd358f47214d1fadddef7f3f62e/clients/node/client-api-gateway-node/types/CreateBasePathMappingInput.ts#L43). Where does it supplied to http handlers?

File: packages/node-http-handler/src/node-http2-handler.ts (hunk 1 above)
Comment (trivikr) [reply packages/node-http-handler/src/node-http2-handler.ts]: I'd created NodeHttp2Options to create sessionTimeout variable.
But reused socketTimeout instead, as both of them mean the same ([documentation](https://nodejs.org/api/http2.html#http2_http2session_and_sockets))

http2 doesn't have keepAlive though

File: packages/node-http-handler/src/node-http2-handler.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      req.on("response", headers => {
+        const httpResponse: HttpResponse<Readable> = {
+          statusCode: headers[":status"] || -1,
+          headers: getTransformedHeaders(headers),
+          body: req
+        };
+        resolve(httpResponse);
+      });
+
+      req.on("error", reject);
```
Comment (AllanFly120) [inline packages/node-http-handler/src/node-http2-handler.ts]: We might need to handle `abort` event: https://nodejs.org/docs/latest/api/http2.html#http2_event_aborted
Same as [`frameError`](https://nodejs.org/docs/latest/api/http2.html#http2_event_frameerror_1), [`trailers`](https://nodejs.org/docs/latest/api/http2.html#http2_event_trailers), [`wantTrailers`](https://nodejs.org/docs/latest/api/http2.html#http2_event_wanttrailers)

File: packages/node-http-handler/src/node-http2-handler.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+      // create the http2 request
+      const req = this.getSession(
+        `${protocol}//${hostname}${port ? `:${port}` : ""}`
+      ).request({
+        ...request.headers,
+        [constants.HTTP2_HEADER_PATH]: queryString
+          ? `${path}?${queryString}`
+          : path,
+        [constants.HTTP2_HEADER_METHOD]: method
```
Comment (AllanFly120) [inline packages/node-http-handler/src/node-http2-handler.ts]: According to the [doc](https://nodejs.org/docs/latest/api/http2.html#http2_headers_object), there are other headers available(`:authority`, `:scheme`). These information should have been specified when you create a connection. So what are they used for? What if these headers are conflict with then endpoint you used to create a connection?

File: packages/node-http-handler/src/node-http2-handler.ts (hunk 1 above)
Comment (trivikr) [reply packages/node-http-handler/src/node-http2-handler.ts]: Resolved in 480dd1bee93c725061661faf7bc6af2c42ed099c

File: packages/node-http-handler/src/node-http2-handler.ts (hunk 3 above)
Comment (trivikr) [reply packages/node-http-handler/src/node-http2-handler.ts]: Added listeners for aborted and frameError events in c49ce80fa130c7520cce4d8c5072938da2d1606d

**Comment by trivikr** [comment]
As TravisCI times out for this PR as described in #346, here's [coverage report](https://github.com/aws/aws-sdk-js-v3/files/3539313/lcov-report.zip) of `@aws-sdk/node-http-handler`

**Comment by trivikr** [comment]
Followed up in PR https://github.com/aws/aws-sdk-js-v3/pull/414 for `smithy-codegen`

**Pull request closed without merging**